### PR TITLE
Unify vkb::Subpass and vkb::rendering::HPPSubpass into vkb::rendering::Subpass<bindingType>

### PR DESCRIPTION
**Pull request by asuessenbach**
## Description

Build tested on Win10 with VS2022. Run tested on Win10 with NVidia GPU.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [x] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [x] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

### CODE REVIEW

**Comment by SaschaWillems** [comment]
This does work fine for me (tested a few samples with the batch run), but I',m kinda unsure about the actual interface.

Once we have unified the repo will we have to add the binding type to each and every class from the framework? If so, I fear that this will make the code hard to read.

e.g. comparing this:
```cpp
std::vector<std::unique_ptr<vkb::Subpass>> subpasses{};
```
with:
```cpp
std::vector<std::unique_ptr<vkb::rendering::Subpass<vkb::BindingType::C>>> subpasses{};
```

Imo readability suffers a lot. Is it possible to somehow omit the binding type unless one explicitly wants to use the CPP variants?

**Comment by asuessenbach** [comment]
I agree, the resulting types are somewhat longer than before.
- I moved the class Subpass from namespace `vkb` to `vkb::rendering`, to make it clear where it resides. If it's consensus that that is a bad move, it would be easy to revert that.
- In order to have the very same class work for both, the C- and the C++-bindings, that BindingType template parameter is added. It is needed. One could, though, add some type aliases like
```
using SubpassC   = Subpass<vkb::BindingType::C>;
using SubpassCpp = Subpass<vkb::BindingType::Cpp>;
```
which would reduce the aforementioned complexity a bit:
`std::vector<std::unique_ptr<vkb::rendering::SubpassC>> subpasses{};`
If it's consensus, I could easily add something like that.

**Review by gpx1000 — DISMISSED** [review]

File: framework/core/command_buffer.cpp [hunk 1]
```diff
@@ -131,7 +131,11 @@ void CommandBuffer::flush(VkPipelineBindPoint pipeline_bind_point)
 	flush_descriptor_state(pipeline_bind_point);
 }
 
-void CommandBuffer::begin_render_pass(const RenderTarget &render_target, const std::vector<LoadStoreInfo> &load_store_infos, const std::vector<VkClearValue> &clear_values, const std::vector<std::unique_ptr<Subpass>> &subpasses, VkSubpassContents contents)
+void CommandBuffer::begin_render_pass(const RenderTarget                                           &render_target,
+                                      const std::vector<LoadStoreInfo>                             &load_store_infos,
+                                      const std::vector<VkClearValue>                              &clear_values,
+                                      const std::vector<std::unique_ptr<vkb::rendering::SubpassC>> &subpasses,
```
Comment (gpx1000) [inline framework/core/command_buffer.cpp:137]: FWIW, some compilers will have issues with >> next to each other.  I don't know if it's been fixed in all compilers or not as I just got into a habit of putting a space between the two.  This compiles and works in Linux; and CI passes so this is probably a non-issue.

File: framework/rendering/subpass.h [hunk 2]
```diff
@@ -40,217 +50,442 @@ struct alignas(16) Light
...
+	using RenderContextType = typename std::conditional<bindingType == vkb::BindingType::Cpp, vkb::rendering::HPPRenderContext, vkb::RenderContext>::type;
+	using RenderTargetType  = typename std::conditional<bindingType == vkb::BindingType::Cpp, vkb::rendering::HPPRenderTarget, vkb::RenderTarget>::type;
 
-	virtual ~Subpass() = default;
+  public:
+	Subpass(RenderContextType &render_context, ShaderSource &&vertex_shader, ShaderSource &&fragment_shader);
 
+	Subpass(const Subpass &)            = delete;
+	Subpass(Subpass &&)                 = default;
+	virtual ~Subpass()                  = default;
```
Comment (gpx1000) [inline framework/rendering/subpass.h:101]: Shouldn't this be override/final instead of virtual here?

File: framework/rendering/subpass.h [hunk 3]
```diff
@@ -40,217 +50,442 @@ struct alignas(16) Light
...
+inline const ShaderSource &Subpass<bindingType>::get_fragment_shader() const
+{
+	return fragment_shader;
+}
 
-	DepthStencilState depth_stencil_state{};
+template <vkb::BindingType bindingType>
+inline void Subpass<bindingType>::set_color_resolve_attachments(std::vector<uint32_t> color_resolve)
+{
+	color_resolve_attachments = color_resolve;
```
Comment (gpx1000) [inline framework/rendering/subpass.h:425]: copy a vector on each call.  maybe move is better?

File: framework/rendering/subpass.h [hunk 4]
```diff
@@ -40,217 +50,442 @@ struct alignas(16) Light
...
+		depth_stencil_resolve_mode = static_cast<vk::ResolveModeFlagBits>(mode);
+	}
+}
 
-	/// Default to no color resolve attachments
-	std::vector<uint32_t> color_resolve_attachments = {};
+template <vkb::BindingType bindingType>
+inline void Subpass<bindingType>::set_input_attachments(std::vector<uint32_t> input)
+{
+	input_attachments = input;
```
Comment (gpx1000) [inline framework/rendering/subpass.h:462]: copy vector on each call.  check if move is correct.

File: framework/rendering/subpass.h [hunk 5]
```diff
@@ -40,217 +50,442 @@ struct alignas(16) Light
...
+}
 
-	/// Default to no depth stencil resolve attachment
-	uint32_t depth_stencil_resolve_attachment{VK_ATTACHMENT_UNUSED};
-};
+template <vkb::BindingType bindingType>
+inline void Subpass<bindingType>::set_output_attachments(std::vector<uint32_t> output)
+{
+	output_attachments = output;
```
Comment (gpx1000) [inline framework/rendering/subpass.h:468]: copy vector on each call.  check if move is correct.

File: framework/core/command_buffer.cpp [hunk 6]
```diff
@@ -277,7 +281,7 @@ void CommandBuffer::bind_index_buffer(const core::Buffer &buffer, VkDeviceSize o
 	vkCmdBindIndexBuffer(get_handle(), buffer.get_handle(), offset, index_type);
 }
 
-void CommandBuffer::bind_lighting(LightingState &lighting_state, uint32_t set, uint32_t binding)
+void CommandBuffer::bind_lighting(vkb::rendering::LightingStateC &lighting_state, uint32_t set, uint32_t binding)
```
Comment (gpx1000) [inline framework/core/command_buffer.cpp:284]: we're already in the vkb namespace.  It's safe to use rendering::LightingStateC

File: framework/core/command_buffer.cpp (hunk 1 above)
Comment (asuessenbach) [reply framework/core/command_buffer.cpp:137]: This was resolved with C++11!
I can't imagine that there's a C++11 compiler out in the wild, that does not support those double closing angle brackets.

File: framework/rendering/subpass.h (hunk 3 above)
Comment (asuessenbach) [reply framework/rendering/subpass.h:425]: Good catch!
Will make it a `std::vector<uint32_t> const &`

File: framework/rendering/subpass.h (hunk 4 above)
Comment (asuessenbach) [reply framework/rendering/subpass.h:462]: As above.

File: framework/rendering/subpass.h (hunk 5 above)
Comment (asuessenbach) [reply framework/rendering/subpass.h:468]: As above.

File: framework/core/command_buffer.cpp (hunk 6 above)
Comment (asuessenbach) [reply framework/core/command_buffer.cpp:284]: I tend to use the complete namespace. I think, it's easier to mentally locate the used entity.
If you want, I can remove that `vbk::` here, though.

**Review by gpx1000 — APPROVED** [review]

**Review by gary-sweet — APPROVED** [review]

**Review by SaschaWillems — APPROVED** [review]
Compiles and runs fine for me. Code changes look fine to me 👍🏻

**Comment by marty-johnson59** [comment]
Merging with 3 approvals.  Thanks!

**Pull request merged**